[Blog] add blogs on running playwright tests in CI using github actions (#162)
JankariTech/blog: src/assets/githubAction/runPlaywrightOnCI.md

---
title: Playwright and GitHub Actions - Run tests in CI
authorName: Nabin Ale
authorAvatar : https://avatars.githubusercontent.com/u/61624650?v=4
authorLink: https://github.com/nabim777
createdAt: Apr 8, 2025
tags: continuous integration, continuous delivery, continuous deployment, github actions, ci, cd, playwright
banner: https://blog.jankaritech.com/src/assets/RunPlaywrightOnCI/images/githubaction_with_playwright_banner
---

This is a blog about how to run Playwright UI tests in GitHub Actions. If you are not familiar and new to UI testing using Playwright, then it would be good to check out this blog about [Playwright](https://blog.jankaritech.com/#/blog/E2E%20Testing%20using%20BDD%20with%20Playwright/Behavior%20Driven%20Development%20(BDD)%20using%20Playwright). Similarly, if you are unfamiliar with GitHub Actions, you can check another blog about [Introduction to GitHub Actions](https://blog.jankaritech.com/#/blog/Introduction%20to%20GitHub%20Actions%20-%20CI%20%26%20CD).

## 🤔 Why to run tests on CI?

> Tests are run on CI(continuous integration) to ensure the code works properly in a clean and isolated environment every time a change is made. Here are some reasons:

Reviewer: ```suggestion
Tests are run on CI(Continuous Integration) to ensure the code works properly in a clean and an isolated environment every time a change is made. Here are some reasons:
```

**1. Early bug detection:**
By automatically running tests after every code commit, you can quickly identify issues as they arise, preventing them from accumulating and causing larger problems later on.

> **2. Fast feedback loop:**

Reviewer 2: how is that different from point 1?

Developers get immediate notification for failing tests. That way they can fix bugs immediately and iterate quickly.

**3. Consistent testing environment:**

> The CI servers run tests in an uniform environment. Therefore, there is no correlation between the developer configurations.

Reviewer 2: I don't understand that point.
correlation between what and what?

**4. Improved code quality:**

> Regular testing done in CI, ensure existing functionality does not become unusable if changes are made.

Reviewer: ```suggestion
Regular testing done in CI, ensures existing functionality does not become unusable if changes are made.
```

> Regular running tests in CI ensures existing functionality does not become unusable if changes are made.

Reviewer 2: for me code quality and functionality are two different things.
Code quality is about having nice structured code, following formatting rules, principles like SOLID etc. Running UI tests in CI will not help with that.
Functionality can be achieved even with bad code quality (even harder and more difficult to maintain). UI tests will only check the functionality, not the code quality.

**5. Reduced integration issues:**

> By regularly integrating and testing the code in CI, you reduce the risk of conflicts when merging large patches of code.

Reviewer 2: how does running the tests in CI help with merge conflicts?
Bob writes some code and adjusts tests => CI is green for him
Alice changes some code in a similar area and adjusts tests => CI is green for her
whoever merges first wins and the second one has to fix the merge-conflict. How does CI help with that?

**6. Automated process:**
CI systems enable developers to save time and effort testing things.

## 📘 About the Project

![Login page of the Project](/src/assets/githubAction/images/project_login_page.png "momo application")

In this blog, I have taken simple applications built using Vue.js. The GitHub repository is available at: https://github.com/nabim777/momo-restro-list.git

> This is a basic application that includes login and logout functionality. E2E test is written using Playwright to verify the login feature.

Reviewer 2: is there only one test or multiple?

> This is a basic application that includes login and logout functionality. A UI test has been written using Playwright to verify the login feature.

Reviewer: ```suggestion
This is a basic application that includes login and logout functionality. E2E test is written using Playwright to verify the login feature.
```

## 🛠️ Running App Locally
Before setting up CI, let's run the application and tests locally.

**1.  Install dependencies**

```bash
npm install
```

**2. Start the frontend and backend servers**

```bash
npm run serve              # Start frontend
npm run backend            # Start backend
```

## 🧪 Running UI Tests Locally

> **_NOTE:_** Make sure both the frontend and backend are running before running the tests.

To run the UI tests locally, use the following command:

```bash
npm run test:e2e tests
```

## ⚙️ Setting Up CI in GitHub Actions
After verifying the app locally, the next step is to set up CI using GitHub Actions. First, create a file named `ci.yml` in a project using the following folder structure:

```
📦momo-restro-list
┗ 📂.github
┃ ┣ 📂workflows
┃ ┃ ┗ 📜ci.yml
```

Then, add the following code in the `ci.yml`.

```yml
name: Run-project

on:
  push:
    branches:
      - master
  pull_request:
    branches:
      - master

jobs:
  run-Restro-project:
    runs-on: ubuntu-latest
    steps:
      - name: Checkout repo code
        uses: actions/checkout@v3

      - name: Set up node
        uses: actions/setup-node@v3
        with:
          node-version: 20.x

>       - name: Run the project
>         run: |
>           npm run serve &
>           npm run backend &

Reviewer 3: ```suggestion
      - name: Install dependencies
        run: npm install
        
      - name: Run the project
        run: |
          npm run serve &
          npm run backend &
```

      - name: Wait for services
        run: |
          sudo apt-get install wait-for-it -y
          wait-for-it -h localhost -p 8080 -t 10
          wait-for-it -h localhost -p 3000 -t 10

      - name: Install playwright and run web-ui tests
        run: |
          npx playwright install
          npm run test:e2e tests
```

## 🔍 What this Workflow Does?

> This GitHub Actions file runs when you push to the `master` branch or create a pull request to `master`

Reviewer 3: ```suggestion
This GitHub Actions file runs when you push to the `master` branch or create a pull request to `master`.
```

This is a basic application that includes login and logout functionality. E2E test is written using Playwright to verify the login feature. You can find the tests scenarios [here](https://github.com/nabim777/momo-restro-list/blob/master/tests/acceptance/features/login.feature)

## 🛠️ Running App Locally
Before setting up CI, let's run the application and tests locally.

**1.  Install dependencies**

```bash
npm install
```

**2. Start the frontend and backend servers**

```bash
npm run serve              # Start frontend
npm run backend            # Start backend
```

## 🧪 Running UI Tests Locally

> **_NOTE:_** Make sure both the frontend and backend are running before running the tests.

To run the UI tests locally, use the following command:

```bash
npm run test:e2e tests
```

## ⚙️ Setting Up CI in GitHub Actions
After verifying the app locally, the next step is to set up CI using GitHub Actions. First, create a file named `ci.yml` in a project using the following folder structure:

```
📦momo-restro-list
┗ 📂.github
┃ ┣ 📂workflows
┃ ┃ ┗ 📜ci.yml
```

Then, add the following code in the `ci.yml`.

```yml

> name: Run-project

Author: The name ` Run UI Tests` is the same as the step name you suggested below.
So, I updated the name with `UI`. If this doesn’t look good, please feel free to suggest alternative names.

on:
  push:
    branches:
      - master
  pull_request:
    branches:
      - master
  workflow_dispatch:

jobs:
  run-Restro-project:
    runs-on: ubuntu-latest
    steps:
      - name: Checkout repo code
        uses: actions/checkout@v3

      - name: Set up node
        uses: actions/setup-node@v3
        with:
          node-version: 20.x

      - name: Install dependencies
        run: |
          npm ci
          npx playwright install chromium

      - name: JS lint
        run: |
          npm run lint || (echo "Linting failed! Please run 'npm run lint:fix' to fix the errors." && exit 1)

      - name: Run the project
        run: |
          npm run serve &
          npm run backend &

      - name: Wait for services
        run: |
          sudo apt-get install wait-for-it -y
          wait-for-it -h localhost -p 8080 -t 10
          wait-for-it -h localhost -p 3000 -t 10

>       - name: Run ui tests

Reviewer 3: ```suggestion
      - name: Run UI tests
```

---
title: Playwright and GitHub Actions - Run tests in CI
authorName: Nabin Ale
authorAvatar : https://avatars.githubusercontent.com/u/61624650?v=4
authorLink: https://github.com/nabim777

> createdAt: Apr 8, 2025

Reviewer 2: should be changed before merging

Author: Okay 👍🏻

---

This blog post explains how to run Playwright UI tests using GitHub Actions.
Before you start, you need to have a basic knowledge of **GitHub Actions**, **Playwright**, and the **Playwright Trace Viewer** to better understand this blog. If you are not familiar with any of these topics, below are links to our blogs that explain each one:
1. [Introduction to GitHub Actions](https://blog.jankaritech.com/#/blog/Introduction%20to%20GitHub%20Actions%20-%20CI%20%26%20CD)
2. [Playwright](https://blog.jankaritech.com/#/blog/E2E%20Testing%20using%20BDD%20with%20Playwright/Behavior%20Driven%20Development%20(BDD)%20using%20Playwright)
3. [Debugging and Error Tracing in Playwright](https://blog.jankaritech.com/#/blog/E2E%20Testing%20using%20BDD%20with%20Playwright/Debugging%20and%20Error%20Tracing%20in%20Playwright)

By the end of this blog, you will be able to set up a CI workflow that automatically runs your Playwright UI tests whenever you push code to your repository. You will also learn how to get Playwright trace reports when tests fail, making it easier to debug and fix errors.

## 🤔 Why run tests on CI?

> Tests are run on CI (continuous integration) to ensure the code works properly in a clean and isolated environment every time a change is made. Here are some reasons:

Reviewer: ```suggestion
Tests are run on CI (Continuous Integration) to ensure the code works properly in a clean and isolated environment every time a change is made. Here are some reasons:
```

> Tests are run on CI (continuous integration) to ensure the code works properly in a clean and isolated environment every time a change is made. Here are some reasons:

Reviewer: You can mention the full form of CI in the first use only then you can just use abbreviation in later uses.

        uses: actions/upload-artifact@v6
        with:
          path: |
            trace-results/*.zip
            retention-days: 30
```

## 🔍 What does this workflow do?
This GitHub Actions file runs when you push to the `master` branch or create a pull request to `master`.

It has one job called `e2e-UI-tests` with these steps:
1. **Checkout repo code** - Gets the project code from GitHub.
2. **Set up node** - Installs Node.js version 24.
3. **Install dependencies** - Installs the project dependencies and Playwright browsers.
4. **JS lint** - Runs the linter to check for code quality issues. If linting fails, it will print a custom error message and exit with a failure status.
5. **Run the project** - Starts the Vue app and the backend using json-server.
6. **Wait for services** - Waits for the frontend (port 8080) and backend (port 3000) to be ready.
7. **Run ui tests** - Runs the UI tests using Playwright.
8. **Upload trace results** - If the UI tests fail, it uploads the Playwright trace results as an artifact that can be downloaded from the GitHub Actions interface. The trace files will be retained for 30 days.

> ## 📥 How to Download and View Trace Files

Reviewer 4: ```suggestion
## 📥 How to Download and View Trace Files?
```
